src/db/mysql/models.rs

```diff
@@ -606,23 +623,28 @@ impl MysqlDb {
             query = query.filter(bso::id.eq_any(ids));
         }
 
+        if let Some(offset_v) = offset {
+            if let Some(offset_id) = offset_v.offset {
+                query = match sort {
+                    Sorting::Oldest => query.filter(bso::id.ge(offset_id)),
+                    _ => query.filter(bso::id.le(offset_id)),
+                }
+            }
+        }
+
         query = match sort {
             Sorting::Index => query.order(bso::sortindex.desc()),
-            Sorting::Newest => query.order(bso::modified.desc()),
-            Sorting::Oldest => query.order(bso::modified.asc()),
-            _ => query,
+            Sorting::Newest | Sorting::None => query.order((bso::modified.desc(), bso::id.desc())),
+            Sorting::Oldest => query.order((bso::modified.asc(), bso::id.asc())),
         };
 
-        // negative limits are no longer allowed by mysql.
-        let limit = limit.map(i64::from).unwrap_or(DEFAULT_LIMIT as i64).max(0);
+        // it is possible to request a limit of 0. use '-1' to indicate no limit, then
+        // don't actually specify a limit later.
+        let limit = limit.map(i64::from).unwrap_or(-1);
         // fetch an extra row to detect if there are more rows that
-        // match the query conditions. Negative limits will cause an error.
-        query = query.limit(if limit == 0 { limit } else { limit + 1 });
-        let numeric_offset = offset.map_or(0, |offset| offset.offset as i64);
-        if numeric_offset != 0 {
-            // XXX: copy over this optimization:
-            // https://github.com/mozilla-services/server-syncstorage/blob/a0f8117/syncstorage/storage/sql/__init__.py#L404
-            query = query.offset(numeric_offset);
+        // match the query conditions. Negative limits will cause an error in mysql.
+        if limit >= 0 {
```

Comment:
Was the negative limit equivalent to specifying no limit?

Reply:
Yeah, it was. The code originally passed it to the limit, which I'm not quite sure did the right thing in any case. I did a first run band-aid where I set the limit to something huge, but then realized a way to just not add the limit if it's not specified.